### PR TITLE
refactor(apiserver): disable insecure port

### LINKED ISSUE #91506: Remove the apiserver insecure port
The API server's "insecure-port" has been [deprecated since 1.10](https://github.com/kubernetes/kubernetes/commit/2f175bc43279a1a4552610a42af0a006f4c6fba2), but never had a deadline for removal. Previous efforts to remove the port have been blocked by requirements for http healthz checks (https://github.com/kubernetes/kubernetes/issues/43784).

I propose publishing a deadline to remove the feature in v1.20. This means that in v1.19 we:
1. Update the deprecation notice to indicate that removal will happen in v1.20 (rather than a "future release")
2. Add an [action required] release note to highlight the removal
3. Figure out a path forward to meeting healthz or other requriments

For item 3, there are already out-of-tree solutions available today, such as a healthz-only proxy which can serve the same API on the same port, and authenticates as a user that only has access to healthz.

/cc @justinsb @micahhausler @deads2k @liggitt 
/sig auth api-machinery
/milestone v1.20
/kind cleanup
/priority important-longterm

### PR DESCRIPTION
**Pull request by knight42**
**What type of PR is this?**

/kind cleanup
/kind deprecation

**What this PR does / why we need it**:

As @liggitt suggests https://github.com/kubernetes/kubernetes/issues/91506#issuecomment-636112285:
> Also, I would recommend leaving the inert flag in place for a few releases even after the insecure serving code is removed. I want to avoid accidents where people stop setting --insecure-port=0 too early in response to warnings about the flag being removed, and accidentally opt back into the default behavior of enabling the insecure port.

I decide to leave the insecure flags intact but make apiserver stop serving on insecure port.

**Which issue(s) this PR fixes**:

xref: #91506

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note-action-required
ACTION REQUIRED: The kube-apiserver ability to serve on an insecure port, deprecated since v1.10, has been removed. The insecure address flags `--address` and `--insecure-bind-address` have no effect in kube-apiserver and will be removed in v1.24. The insecure port flags `--port` and `--insecure-port` may only be set to 0 and will be removed in v1.24.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
> It might be better to disable insecure port by default first, and then completely remove it after a few releases.

I don't think we want a multi-stage rollout of this... that will lead people to flip their invocations to explicitly enable the insecure port, which is not desired.

**Comment by dims** [comment]
+1 to what @liggitt said!

**Comment by knight42** [comment]
@liggitt @dims I made apiserver stop serving on insecure port while leaving the insecure flags intact, how is that sound to you?

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/95522 was in progress to remove insecure serving ability... I expected the insecure flags to be registered to dummy variables with deprecated warnings as part of that PR

**Comment by knight42** [comment]
@liggitt Oh I didn't notice #95522.. I have updated the code to register insecure flags to dummy variables. PTAL

**Comment by cheftako** [comment]
Note: Deprecated 1.12, 2018 in https://github.com/kubernetes/kubernetes/pull/67060

**Comment by fedebongio** [comment]
/cc @deads2k @lavalamp 
to keep this one in your radar
/triage accepted

**Comment by knight42** [comment]
@liggitt All tests passed now, PTAL.

Note: since apiserver no longer serves on insecure port, we have to authenticate with apiserver in tests and I decided to use the admin token here, so it is mandatory to pass `$kube_flags` to kubectl in `test/cmd/*.sh` now.

**Comment by liggitt** [comment]
/cc @mfojtik 

not sure how this overlaps with #95522

**Comment by knight42** [comment]
> not sure how this overlaps with #95522

@liggitt  IIUC this PR only disable insecure port in apiserver, while #95522 removes insecure serving from apiserver, controller-manager and schduler which would require too much change in the test. I guess we'd better remove insecure serving step by step.

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -134,14 +134,35 @@ func NewServerRunOptions() *ServerRunOptions {
...
+func addDummyInsecureFlags(fs *pflag.FlagSet) {
+	var (
+		bindAddr = net.IPv4(127, 0, 0, 1)
+		bindPort = 8080
+	)
+
+	for _, name := range []string{"insecure-bind-address", "address"} {
+		fs.IPVar(&bindAddr, name, bindAddr, ""+
+			"The IP address on which to serve the insecure port (set to 0.0.0.0 for all IPv4 interfaces and :: for all IPv6 interfaces).")
+		// Though this flag is deprecated, we discovered security concerns over how to do health checks without it e.g. #43784
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:146]: drop this line

File: cmd/kube-apiserver/app/options/options.go [hunk 2]
```diff
@@ -134,14 +134,35 @@ func NewServerRunOptions() *ServerRunOptions {
...
+		fs.MarkDeprecated(name, "This flag has no effect now and will be removed in a future version.")
+		fs.Lookup(name).Hidden = false
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:148]: remove this line, we can hide the flag now

File: cmd/kube-apiserver/app/options/options.go [hunk 3]
```diff
@@ -134,14 +134,35 @@ func NewServerRunOptions() *ServerRunOptions {
...
+	}
+
+	for _, name := range []string{"insecure-port", "port"} {
+		fs.IntVar(&bindPort, name, bindPort, ""+
+			"The port on which to serve unsecured, unauthenticated access.")
+		fs.MarkDeprecated(name, "This flag has no effect now and will be removed in a future version.")
+		fs.Lookup(name).Hidden = false
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:155]: remove this line, we can hide the flag now

File: cmd/kube-apiserver/app/options/options.go [hunk 4]
```diff
@@ -134,14 +134,35 @@ func NewServerRunOptions() *ServerRunOptions {
 	return &s
 }
 
+func addDummyInsecureFlags(fs *pflag.FlagSet) {
+	var (
+		bindAddr = net.IPv4(127, 0, 0, 1)
+		bindPort = 8080
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:140]: since this is inert, the behavior is equivalent to a default of 0 now

File: hack/make-rules/test-cmd.sh [hunk 5]
```diff
@@ -73,7 +71,7 @@ function run_kube_apiserver() {
     --token-auth-file=hack/testdata/auth-tokens.csv 1>&2 &
   export APISERVER_PID=$!
 
-  kube::util::wait_for_url "http://127.0.0.1:${API_PORT}/healthz" "apiserver"
+  kube::util::wait_for_url "https://127.0.0.1:${SECURE_API_PORT}/healthz" "apiserver" 1 30 1 '--oauth2-bearer admin-token'
```
Comment (liggitt) [inline hack/make-rules/test-cmd.sh:74]: changing `AUTHORIZATION_MODE` to `"RBAC,Node"` would enable anonymous auth and allow healthz to be fetched without passing these extra args

File: hack/update-openapi-spec.sh [hunk 6]
```diff
@@ -50,33 +50,33 @@ kube::golang::setup_env
...
   --bind-address="${API_HOST}" \
-  --insecure-port="${API_PORT}" \
+  --secure-port="${API_PORT}" \
   --etcd-servers="http://${ETCD_HOST}:${ETCD_PORT}" \
   --advertise-address="10.10.10.10" \
   --cert-dir="${TMP_DIR}/certs" \
   --runtime-config="api/all=true" \
   --token-auth-file="${TMP_DIR}/tokenauth.csv" \
-  --service-account-issuer="https://kubernetes.devault.svc/" \
+  --service-account-issuer="https://kubernetes.default.svc/" \
```
Comment (liggitt) [inline hack/update-openapi-spec.sh:72]: add `--authorization-mode=RBAC`, which will allow anonymous access to healthz

File: hack/update-openapi-spec.sh [hunk 7]
```diff
@@ -50,33 +50,33 @@ kube::golang::setup_env
...
   --service-account-signing-key-file="${KUBE_ROOT}/staging/src/k8s.io/client-go/util/cert/testdata/dontUseThisKey.pem" \
   --logtostderr \
   --v=2 \
   --service-cluster-ip-range="10.0.0.0/24" >"${API_LOGFILE}" 2>&1 &
 APISERVER_PID=$!
 
-if ! kube::util::wait_for_url "${API_HOST}:${API_PORT}/healthz" "apiserver: "; then
+if ! kube::util::wait_for_url "https://${API_HOST}:${API_PORT}/healthz" "apiserver: " 1 30 1 "--oauth2-bearer ${AUTH_TOKEN}"; then
   kube::log::error "Here are the last 10 lines from kube-apiserver (${API_LOGFILE})"
   kube::log::error "=== BEGIN OF LOG ==="
```
Comment (liggitt) [inline hack/update-openapi-spec.sh:81]: drop this change

File: hack/lib/util.sh [hunk 8]
```diff
@@ -54,7 +55,9 @@ kube::util::wait_for_url() {
   local i
   for i in $(seq 1 "${times}"); do
     local out
-    if out=$(curl --max-time "${maxtime}" -gkfs "${url}" 2>/dev/null); then
+    # $extra_args might contain multiple flags, we do want to split words here.
+    # shellcheck disable=SC2086
+    if out=$(curl --max-time "${maxtime}" -gkfs $extra_args "${url}" 2>/dev/null); then
```
Comment (liggitt) [inline hack/lib/util.sh:60]: drop this change, switching away from AlwaysAllow authorizer will remove the need to pass bearer tokens to check healthz

File: hack/make-rules/test-cmd.sh [hunk 9]
```diff
@@ -101,7 +121,7 @@ function run_kube_controller_manager() {
 # Exports:
 #   SUPPORTED_RESOURCES(Array of all resources supported by the apiserver).
 function create_node() {
-  kubectl create -f - -s "http://127.0.0.1:${API_PORT}" << __EOF__
+  kubectl create -s "https://127.0.0.1:${SECURE_API_PORT}" --insecure-skip-tls-verify --token=admin-token -f - << __EOF__
```
Comment (liggitt) [inline hack/make-rules/test-cmd.sh:124]: I expected to set $KUBECONFIG pointing at a file containing credentials so we don't have to set the token in the invocation

File: hack/update-openapi-spec.sh [hunk 10]
```diff
@@ -50,33 +50,33 @@ kube::golang::setup_env
 TMP_DIR=$(mktemp -d /tmp/update-openapi-spec.XXXX)
 ETCD_HOST=${ETCD_HOST:-127.0.0.1}
 ETCD_PORT=${ETCD_PORT:-2379}
-API_PORT=${API_PORT:-8050}
+API_PORT=${API_PORT:-6443}
```
Comment (liggitt) [inline hack/update-openapi-spec.sh:53]: why was this changed?

File: test/cmd/apply.sh [hunk 11]
```diff
@@ -94,7 +94,7 @@ run_kubectl_apply_tests() {
   output_message=$(! kubectl get pods selector-test-pod-dont-apply 2>&1 "${kube_flags[@]:?}")
   kube::test::if_has_string "${output_message}" 'pods "selector-test-pod-dont-apply" not found'
   # cleanup
-  kubectl delete pods selector-test-pod
+  kubectl delete "${kube_flags[@]:?}" pods selector-test-pod
```
Comment (liggitt) [inline test/cmd/apply.sh:97]: global comment - I expected to set $KUBECONFIG pointing at a file containing credentials so we don't have to set the token in the invocation

**Comment by liggitt** [comment]
> @liggitt IIUC this PR only disable insecure port in apiserver, while #95522 removes insecure serving from apiserver, controller-manager and schduler which would require too much change in tests. I guess we'd better remove insecure serving step by step.

I think that makes sense. fyi @mfojtik

File: hack/update-openapi-spec.sh (hunk 10 above)
Comment (knight42) [reply hack/update-openapi-spec.sh:53]: I think 6443 indicates this is a secure port, I could revert this change if you want.

Comment (liggitt) [reply hack/update-openapi-spec.sh:53]: yeah, I think the goal was to pick a port that would not conflict with an existing running server

File: test/cmd/apply.sh (hunk 11 above)
Comment (knight42) [reply test/cmd/apply.sh:97]: This is a good point, thanks for pointing it out! since its already midnight here, I would make this change tomorrow.

File: hack/make-rules/test-cmd.sh (hunk 5 above)
Comment (knight42) [reply hack/make-rules/test-cmd.sh]: If we changed `AUTHORIZATION_MODE` from `RBAC,AlwaysAllow` to `RBAC,Node`, the test [`run_impersonation_tests`](https://github.com/kubernetes/kubernetes/blob/518b826b1dcc4852226300ae08110cba03a078c9/test/cmd/authorization.sh#L56) would not pass, because it needs to run `kubectl create ... --as=user1`.

I could think of 2 solutions:
1. leave `AUTHORIZATION_MODE` intact, and add a new util function `kube::util::wait_for_url_with_bearer_token` which accesses the url with given token.
2. change `AUTHORIZATION_MODE` to `RBAC,Node`, and add a new token whose user name is `user1` to https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/auth-tokens.csv

I decided to use 1) here, WDYT?

File: cmd/kube-apiserver/app/options/options.go [hunk 12]
```diff
@@ -134,14 +134,32 @@ func NewServerRunOptions() *ServerRunOptions {
 	return &s
 }
 
+func addDummyInsecureFlags(fs *pflag.FlagSet) {
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:137]: I'd probably set v1.24 as a target removal version in the message and in a TODO

File: cmd/kube-apiserver/app/options/options.go [hunk 13]
```diff
@@ -134,14 +134,32 @@ func NewServerRunOptions() *ServerRunOptions {
...
+		fs.IPVar(&bindAddr, name, bindAddr, ""+
+			"The IP address on which to serve the insecure port (set to 0.0.0.0 for all IPv4 interfaces and :: for all IPv6 interfaces).")
+		fs.MarkDeprecated(name, "This flag has no effect now and will be removed in a future version.")
+	}
+
+	for _, name := range []string{"insecure-port", "port"} {
+		fs.IntVar(&bindPort, name, bindPort, ""+
+			"The port on which to serve unsecured, unauthenticated access.")
+		fs.MarkDeprecated(name, "This flag has no effect now and will be removed in a future version.")
+	}
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:153]: the only other thing I think would be good is if we can fail if someone tries to explicitly set the port args to a non-zero value... the reason we're keeping these registered is for people who were explicitly disabling the ports

File: cmd/kube-apiserver/app/options/options.go (hunk 12 above)
Comment (knight42) [reply cmd/kube-apiserver/app/options/options.go:140]: added

File: cmd/kube-apiserver/app/options/options.go (hunk 13 above)
Comment (knight42) [reply cmd/kube-apiserver/app/options/options.go]: The value of these flags is checked now.

**Comment by deads2k** [comment]
I like the change and the approach.  I would also support removing insecure options from kube-controller-manager and kube-scheduler in separate PRs.

/approve

**Comment by liggitt** [comment]
Updated release note

/lgtm

**Comment by neolit123** [comment]
@liggitt i think we should prefix the release note with action required.

**Comment by liggitt** [comment]
weird... I did with `release-note-action-required` but the bot didn't add the right label

**Comment by knight42** [comment]
looks like we have to modify the release note.
